x/bank/keeper/restrictions.go

```diff
@@ -0,0 +1,115 @@
+package keeper
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// A MintingRestrictionFn can restrict minting of coins.
+type MintingRestrictionFn func(ctx sdk.Context, coins sdk.Coins) error
+
+var _ MintingRestrictionFn = NoOpMintingRestrictionFn
+
+// NoOpMintingRestrictionFn is a no-op MintingRestrictionFn.
+func NoOpMintingRestrictionFn(_ sdk.Context, _ sdk.Coins) error {
+	return nil
+}
+
+// Then creates a composite restriction that runs this one then the provided second one.
+func (r MintingRestrictionFn) Then(second MintingRestrictionFn) MintingRestrictionFn {
+	return ComposeMintingRestrictions(r, second)
+}
+
+// ComposeMintingRestrictions combines multiple MintingRestrictionFn into one.
+// nil entries are ignored.
+// If all entries are nil, nil is returned.
+// If exactly one entry is not nil, it is returned.
+// Otherwise, a new MintingRestrictionFn is returned that runs the non-nil restrictions in the order they are given.
+func ComposeMintingRestrictions(restrictions ...MintingRestrictionFn) MintingRestrictionFn {
+	toRun := make([]MintingRestrictionFn, 0, len(restrictions))
+	for _, r := range restrictions {
+		if r != nil {
+			toRun = append(toRun, r)
+		}
+	}
+	switch len(toRun) {
+	case 0:
+		return nil
+	case 1:
+		return toRun[0]
+	}
+	return func(ctx sdk.Context, coins sdk.Coins) error {
+		for _, r := range toRun {
+			err := r(ctx, coins)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// A SendRestrictionFn can restrict sends and/or provide a new receiver address.
+type SendRestrictionFn func(ctx sdk.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) (newToAddr sdk.AccAddress, err error)
+
+var _ SendRestrictionFn = NoOpSendRestrictionFn
+
+// NoOpSendRestrictionFn is a no-op SendRestrictionFn.
+func NoOpSendRestrictionFn(_ sdk.Context, _, toAddr sdk.AccAddress, _ sdk.Coins) (sdk.AccAddress, error) {
+	return toAddr, nil
+}
+
+// Then creates a composite restriction that runs this one then the provided second one.
+func (r SendRestrictionFn) Then(second SendRestrictionFn) SendRestrictionFn {
+	return ComposeSendRestrictions(r, second)
+}
+
+// ComposeSendRestrictions combines multiple SendRestrictionFn into one.
+// nil entries are ignored.
+// If all entries are nil, nil is returned.
+// If exactly one entry is not nil, it is returned.
+// Otherwise, a new SendRestrictionFn is returned that runs the non-nil restrictions in the order they are given.
```

Comment:
Please specify the behavior of the composition too, here and for `Then()`, `Append()`, and `Prepend()` - something like:

> The composition runs each send restriction until an error is encountered and returns that error, otherwise it returns the `toAddr` of the last send restriction.

Reply:
Good call. Added it.